core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -1424,11 +1424,18 @@ class UnifiedLog(@volatile var logStartOffset: Long,
 
   /**
     * Given a message offset, find its corresponding offset metadata in the log.
-    * If the message offset is out of range, throw an OffsetOutOfRangeException
+    * 1. If the message offset is lesser than the log-start-offset, then throw an OffsetOutOfRangeException
```

Comment:
For case 1, it looks like we never throw `OffsetOutOfRangeException` now, doesn't it?

Reply:
`checkLogStartOffset` will throw `OffsetOutOfRangeException` if the offset is lesser than the `logStartOffset`.

Reply:
Should we avoid throwing the error and return message-only metadata when the offset is lesser than the log-start-offset?

While updating the [UnifiedLog#updateLogStartOffset](https://sourcegraph.com/github.com/apache/kafka@5f933ac840343429e911f5706ccb7cd8dc379462/-/blob/core/src/main/scala/kafka/log/UnifiedLog.scala?L535), the HWM also gets updated and it doesn't hit the `fetchHighWatermarkMetadata` (or) `convertToOffsetMetadataOrThrow` so the call will succeed even when log-start-offset > HWM.

Reply:
Ah, OK. Thanks.

Reply:
> While updating the [UnifiedLog#updateLogStartOffset](https://sourcegraph.com/github.com/apache/kafka@5f933ac840343429e911f5706ccb7cd8dc379462/-/blob/core/src/main/scala/kafka/log/UnifiedLog.scala?L535), the HWM also gets updated and it doesn't hit the fetchHighWatermarkMetadata (or) convertToOffsetMetadataOrThrow so the call will succeed even when current-log-start-offset > old-HWM.

For [UnifiedLog#updateLogStartOffset](https://sourcegraph.com/github.com/apache/kafka@5f933ac840343429e911f5706ccb7cd8dc379462/-/blob/core/src/main/scala/kafka/log/UnifiedLog.scala?L535), I think it's safe we didn't throw exception when `current-log-start-offset > old-HWM` because it will be called when initialization or `maybeIncrementLogStartOffset`. In the latter, we've checked `newLogStartOffset > logStartOffset` already.

Reply:
Good point. Since we change the logic such that it's ok not to have the metadata for an offset, we could just skip the call to `checkLogStartOffset`.

Reply:
Removed the `checkLogStartOffset` from the `convertToOffsetMetadataOrThrow` method.